Cubical/Categories/Instances/FPAlgebrasSmall.agda

```diff
@@ -0,0 +1,334 @@
+open import Cubical.Categories.Instances.CommRings
+open import Cubical.Categories.Instances.CommAlgebras
+open import Cubical.Categories.Instances.DistLattices
+open import Cubical.Categories.Instances.Functors
+open import Cubical.Categories.NaturalTransformation
+open import Cubical.Categories.Presheaf.Base
+open import Cubical.Categories.Yoneda
+
+open import Cubical.Relation.Binary.Order.Poset
+
+open import Cubical.HITs.PropositionalTruncation as PT
+open import Cubical.HITs.SetQuotients as SQ
+
+
+private
+ variable
+  ℓ ℓ' ℓ'' : Level
+
+
+module _ (R : CommRing ℓ) where
+
+  open Iso
+  open Category hiding (_∘_) renaming (_⋆_ to _⋆c_)
+  open Functor
+  open NatTrans
+  open DistLatticeStr ⦃...⦄
+  open CommRingStr ⦃...⦄
+  open IsRingHom
+  open CommAlgebraHoms
+  open IsLatticeHom
+  open ZarLat
+  open ZarLatUniversalProp
+
+
+  -- type living in the same universe as the base ring R
+  record SmallFPAlgebra : Type ℓ where
+    field
+      n : ℕ
+      m : ℕ
+      relations : FinVec ⟨ Polynomials {R = R} n ⟩ m
+
+  open SmallFPAlgebra
+  private
+    indFPAlg : SmallFPAlgebra → CommAlgebra R ℓ
+    indFPAlg A = FPAlgebra (A .n) (A .relations)
+
+  FPAlgebrasSmallCategory : Category ℓ ℓ
+  ob FPAlgebrasSmallCategory = SmallFPAlgebra
+  Hom[_,_] FPAlgebrasSmallCategory A B =
+    CommAlgebraHom (indFPAlg A) (indFPAlg B)
+  id FPAlgebrasSmallCategory = idCommAlgebraHom _
+  _⋆c_ FPAlgebrasSmallCategory = compCommAlgebraHom _ _ _
+  ⋆IdL FPAlgebrasSmallCategory = compIdCommAlgebraHom
+  ⋆IdR FPAlgebrasSmallCategory = idCompCommAlgebraHom
+  ⋆Assoc FPAlgebrasSmallCategory = compAssocCommAlgebraHom
+  isSetHom FPAlgebrasSmallCategory = isSetAlgebraHom _ _
+
+
+  -- the name of this module will be justified once we have defined
+  -- Zariski sheaves on small f.p. algebras
+  module SchemesOfFinitePresentation where
+
+    -- "parsimonious" presheaf categories
+    fpAlgFunctor = Functor FPAlgebrasSmallCategory (SET ℓ)
+    fpAlgFUNCTOR = FUNCTOR FPAlgebrasSmallCategory (SET ℓ)
+
+    -- yoneda in the notation of Demazure-Gabriel
+    -- uses η-equality of Categories
+    Sp : Functor (FPAlgebrasSmallCategory ^op) fpAlgFUNCTOR
+    Sp = YO {C = (FPAlgebrasSmallCategory ^op)}
+
+    -- the Zariski lattice (classifying compact open subobjects)
+    private forget = ForgetfulCommAlgebra→CommRing R
+
+    𝓛 : fpAlgFunctor
+    F-ob 𝓛 A = ZL (forget .F-ob (indFPAlg A)) , SQ.squash/
+    F-hom 𝓛  φ = inducedZarLatHom (forget .F-hom φ) .fst
+    F-id 𝓛 = cong (λ x → inducedZarLatHom x .fst) (forget .F-id)
+            ∙ cong fst (inducedZarLatHomId _)
+    F-seq 𝓛 _ _ = cong (λ x → inducedZarLatHom x .fst) (forget .F-seq _ _)
+                 ∙ cong fst (inducedZarLatHomSeq _ _)
+
+    CompactOpen : fpAlgFunctor → Type ℓ
+    CompactOpen X = X ⇒ 𝓛
+
+    isAffine : fpAlgFunctor → Type ℓ
+    isAffine X = ∃[ A ∈ SmallFPAlgebra ] (Sp .F-ob A) ≅ᶜ X
+
+    -- the induced subfunctor
+    ⟦_⟧ᶜᵒ : {X : fpAlgFunctor} (U : CompactOpen X) → fpAlgFunctor
```

Comment:
Maybe this should be defined by just taking the pullback of `U` and the "top element inclusion" of the lattice?